Packet++/src/VrrpLayer.cpp

```diff
@@ -0,0 +1,533 @@
+#define LOG_MODULE PacketLogModuleVrrpLayer
+
+#include <SystemUtils.h>
+#include "PacketUtils.h"
+#include "Logger.h"
+#include "EndianPortable.h"
+#include "IPv4Layer.h"
+#include "PayloadLayer.h"
+#include "VrrpLayer.h"
+
+namespace pcpp {
+
+	/*************
+	 * VrrpLayer
+	 *************/
+
+	VrrpLayer::VrrpLayer(ProtocolType subProtocol) {
+		m_DataLen = VRRP_PACKET_FIX_LEN;
+		m_Data = new uint8_t[m_DataLen];
+		memset(m_Data, 0, m_DataLen);
+		m_Protocol = subProtocol;
+
+		getPacketPtr()->ipAddrCount = 0;
+		getPacketPtr()->version = getVersionFromData();
+
+		m_addressType = IPAddress::IPv4AddressType;
+	}
+
+	bool VrrpLayer::isDataValid(size_t dataLen) {
+		return (dataLen > VRRP_PACKET_FIX_LEN);
+	}
+
+	VrrpLayer *VrrpLayer::parseVrrpLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet,
+										 IPAddress::AddressType addressType) {
+		auto *vrrpPacketCommon = (vrrp_packet *) data;
+		uint8_t version = vrrpPacketCommon->version;
+		switch (version) {
+			case Vrrp_Version_2:
+				return new VrrpV2Layer(data, dataLen, prevLayer, packet);
+			case Vrrp_Version_3:
+				return new VrrpV3Layer(data, dataLen, prevLayer, packet, addressType);
+			default: {
+				return nullptr;
+			}
+		}
+	}
+
+	void VrrpLayer::computeCalculateFields() {
+		// calculate and fill the checksum to packet
+		fillChecksum();
+	}
+
+	uint8_t VrrpLayer::isIPv4Packet() const {
+		return (getAddressType() == IPAddress::IPv4AddressType);
+	}
+
+	uint8_t VrrpLayer::isIPv6Packet() const {
+		return (getAddressType() == IPAddress::IPv6AddressType);
+	}
+
+	uint8_t VrrpLayer::getIPAddressLen() const {
+		if (isIPv4Packet()) {
+			return 4;
+		} else if (isIPv6Packet()) {
+			return 16;
+		} else {
+			return 0;
+		}
+	}
+
+	uint8_t VrrpLayer::getVersionFromData() const {
+		vrrp_packet *packet = getPacketPtr();
+		return (packet->version);
+	}
+
+	void VrrpLayer::setPacket(vrrp_packet *packet) {
+		memcpy(m_Data, packet, sizeof(vrrp_packet));
+	}
+
+	bool VrrpLayer::isChecksumCorrect() const {
+		const vrrp_packet *packet = (vrrp_packet *) getData();
+		if (packet == nullptr) {
+			return false;
+		}
+
+		return (calculateChecksum() == be16toh(packet->checksum));
+	}
+
+	std::string VrrpLayer::getTypeName() const {
+		std::string typeName;
+		if (getPacketPtr()->type == VrrpType_Advertisement) {
+			typeName = "Advertisement";
+		} else {
+			typeName = "Unknown";
+		}
+
+		return typeName;
+	}
+
+	std::string VrrpLayer::getPriorityDesc() const {
+		std::string priorityName;
+		if (getPacketPtr()->priority == VRRP_PRIO_DFF) {
+			priorityName = "(Default priority for a backup VRRP router)";
+		} else if (getPriority() == VRRP_PRIO_STOP) {
+			priorityName = "(Current Master has stopped participating in VRRP)";
+		} else if (getPriority() == VRRP_PRIO_OWNER) {
+			priorityName = "(This VRRP router owns the virtual router's IP address(es))";
+		} else {
+		}
+
+		return priorityName;
+	}
```

Comment:
if we make `toString()` shorter this method is probably not needed

Reply:
The macro define is used here. I add this function for short function.